File: internal/web/static/js/modules/reasoning-effort.test.js
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import vm from 'node:vm';

function load() {
  const window = {};
  const source = readFileSync(new URL('./reasoning-effort.js', import.meta.url), 'utf8');
  vm.runInNewContext(source, { window }, { filename: 'reasoning-effort.js' });
  return window.OriReasoningEffort;
}

const asData = value => JSON.parse(JSON.stringify(value));

function fakeSelect() {
  const doc = {
    createElement: tag => ({ tag, value: '', textContent: '', selected: false })
  };
  return {
    ownerDocument: doc,
    options: [],
    value: 'stale',
    replaceChildren() {
      this.options = [];
    },
    appendChild(option) {
      this.options.push(option);
    }
  };
}

test('only Codex and Claude Code take a reasoning level', () => {
  const api = load();
  assert.deepEqual(asData(api.levels('codex', 'gpt-5.6-sol')), ['low', 'medium', 'high', 'xhigh']);
  assert.deepEqual(asData(api.levels('openai', 'gpt-5-codex')), ['low', 'medium', 'high', 'xhigh']);
  assert.deepEqual(asData(api.levels('claude_code', 'opus')), [
    'low',
    'medium',
    'high',
    'xhigh',
    'max'
  ]);
  for (const [provider, model] of [
    ['claude', 'claude-opus-5'],
    ['openai', 'gpt-5.6-sol'],
    ['ollama', 'llama3.2'],
    ['', '']
  ]) {
    assert.equal(api.supports(provider, model), false, `${provider}/${model}`);
    assert.deepEqual(asData(api.options(provider, model)), []);
    assert.equal(api.helpText(provider, model), '');
  }
});

test('levels normalize per provider, with a Codex-only default', () => {
  const api = load();
  assert.equal(api.normalize('claude_code', 'opus', ' MAX '), 'max');
  assert.equal(api.normalize('codex', 'gpt-5.6-sol', 'max'), '');
  assert.equal(api.normalize('claude', 'claude-opus-5', 'high'), '');
  assert.equal(api.defaultFor('codex', 'gpt-5.6-sol'), 'medium');
  assert.equal(api.defaultFor('claude_code', 'opus'), '');

  assert.deepEqual(
    asData(api.options('codex', 'x')).map(option => option.label),
    ['Low', 'Medium (Recommended)', 'High', 'Extra High']
  );
  assert.deepEqual(asData(api.options('claude_code', 'opus')), [
    { value: '', label: 'Claude Code default' },
    { value: 'low', label: 'Low' },
    { value: 'medium', label: 'Medium' },
    { value: 'high', label: 'High' },
    { value: 'xhigh', label: 'Extra High' },
    { value: 'max', label: 'Max' }
  ]);
});

test('a select keeps an accepted choice across providers and falls back otherwise', () => {
  const api = load();
  const select = fakeSelect();

  assert.equal(api.syncSelect(select, 'claude_code', 'opus', 'max'), 'max');
  assert.equal(select.value, 'max');
  assert.equal(select.options.length, 6);

  // Codex has no max: its default is chosen instead.
  assert.equal(api.syncSelect(select, 'codex', 'gpt-5.6-sol', 'max'), 'medium');
  assert.equal(select.options.length, 4);
  assert.equal(api.syncSelect(select, 'codex', 'gpt-5.6-sol', 'high'), 'high');

  // Claude Code's default is "no flag", kept as an explicit empty choice.
  assert.equal(api.syncSelect(select, 'claude_code', 'opus', ''), '');
  assert.equal(api.syncSelect(select, 'claude_code', 'opus', 'nonsense'), '');
  assert.equal(api.syncSelect(select, 'claude_code', 'opus', 'high'), 'high');

  // A provider without levels empties the select.
  assert.equal(api.syncSelect(select, 'claude', 'claude-opus-5', 'high'), '');
  assert.equal(select.options.length, 0);
});

File: internal/types/reasoning_effort_test.go
package types

import (
	"reflect"
	"testing"
)

func TestReasoningEffortLevels_OnlyCodexAndClaudeCode(t *testing.T) {
	cases := []struct {
		provider, model string
		want            []string
	}{
		{"codex", "gpt-5.6-sol", []string{"low", "medium", "high", "xhigh"}},
		{" CODEX ", "", []string{"low", "medium", "high", "xhigh"}},
		{"openai", "gpt-5-codex", []string{"low", "medium", "high", "xhigh"}},
		{"claude_code", "opus", []string{"low", "medium", "high", "xhigh", "max"}},
		{"claude", "claude-opus-5", nil},
		{"openai", "gpt-5.6-sol", nil},
		{"ollama", "llama3.2", nil},
		{"", "", nil},
	}
	for _, tc := range cases {
		if got := ReasoningEffortLevels(tc.provider, tc.model); !reflect.DeepEqual(got, tc.want) {
			t.Errorf("ReasoningEffortLevels(%q, %q) = %v, want %v", tc.provider, tc.model, got, tc.want)
		}
		if got := SupportsReasoningEffort(tc.provider, tc.model); got != (tc.want != nil) {
			t.Errorf("SupportsReasoningEffort(%q, %q) = %v", tc.provider, tc.model, got)
		}
	}

	// Callers cannot mutate the shared table through a returned slice.
	levels := ReasoningEffortLevels("codex", "")
	levels[0] = "tampered"
	if ReasoningEffortLevels("codex", "")[0] != "low" {
		t.Fatal("ReasoningEffortLevels returned the shared backing array")
	}
}

func TestNormalizeReasoningEffortFor_AcceptsOnlyTheProvidersLevels(t *testing.T) {
	cases := []struct {
		provider, model, effort, want string
	}{
		{"codex", "gpt-5.6-sol", " High ", "high"},
		{"codex", "gpt-5.6-sol", "max", ""},
		{"claude_code", "opus", "max", "max"},
		{"claude_code", "opus", "XHIGH", "xhigh"},
		{"claude_code", "opus", "extreme", ""},
		{"claude", "claude-opus-5", "high", ""},
		{"openai", "gpt-5.6-sol", "high", ""},
	}
	for _, tc := range cases {
		if got := NormalizeReasoningEffortFor(tc.provider, tc.model, tc.effort); got != tc.want {
			t.Errorf("NormalizeReasoningEffortFor(%q, %q, %q) = %q, want %q", tc.provider, tc.model, tc.effort, got, tc.want)
		}
	}
	if got := NormalizeReasoningEffort("MAX"); got != "max" {
		t.Errorf("NormalizeReasoningEffort(MAX) = %q, want max", got)
	}
}

func TestEffectiveReasoningEffort_DefaultsCodexOnly(t *testing.T) {
	cases := []struct {
		name     string
		provider string
		settings Settings
		want     string
	}{
		{"codex unset defaults to medium", "codex", Settings{Model: "gpt-5.6-sol"}, "medium"},
		{"codex explicit", "codex", Settings{Model: "gpt-5.6-sol", ReasoningEffort: "xhigh"}, "xhigh"},
		{"codex cannot use a claude-only level", "codex", Settings{Model: "gpt-5.6-sol", ReasoningEffort: "max"}, "medium"},
		{"claude code unset sends nothing", "claude_code", Settings{Model: "opus"}, ""},
		{"claude code explicit", "claude_code", Settings{Model: "opus", ReasoningEffort: "max"}, "max"},
		{"api providers ignore it", "claude", Settings{Model: "claude-opus-5", ReasoningEffort: "high"}, ""},
	}
	for _, tc := range cases {
		if got := tc.settings.EffectiveReasoningEffort(tc.provider); got != tc.want {
			t.Errorf("%s: EffectiveReasoningEffort(%q) = %q, want %q", tc.name, tc.provider, got, tc.want)
		}
	}
}
